### PR TITLE
7.0 bug 1253052 parent order

### PR DESCRIPTION
**Pull request by gurneyalex**
Port of Cedric Le Brouster's fix for https://bugs.launchpad.net/openobject-server/+bug/1253052 to github.

### CODE REVIEW

File: openerp/osv/orm.py [hunk 1]
```diff
@@ -4178,22 +4178,31 @@ def write(self, cr, user, ids, vals, context=None):
...
-                                (self._table, self._parent_name, parent_order)
-                cr.execute(query, (tuple(ids),))
-            parents_changed = map(operator.itemgetter(0), cr.fetchall())
+            query = "SELECT id FROM %s WHERE id IN %%s" % (self._table,)
+            query_params = [tuple(ids)]
+            query_clause = ""
+            parent_params = [self._parent_name]
+            for param in parent_order.split(','):
+                parent_params.append(param.strip())
+            for parent_param in parent_params:
```
Comment (rco-odoo) [inline openerp/osv/orm.py]: Variable name is confusing: `parent_params` is not parameters, they are field names; so please use a better name like `field_names` or something like that.

File: openerp/osv/orm.py [hunk 2]
```diff
@@ -4178,22 +4178,31 @@ def write(self, cr, user, ids, vals, context=None):
...
+                if parent_param in vals:
+                    parents_changed = True
+                    parent_val = vals[parent_param]
+                    if parent_val:
+                        query_clause += (query_clause and ' OR ' or '') + "%s != %%s OR %s IS NULL" % (parent_param,parent_param)
```
Comment (rco-odoo) [inline openerp/osv/orm.py]: Please avoid complex string manipulations. Simply use a list `query_clauses`, append the clauses to it, and join them with `' OR '.join(query_clauses)` into the query.

File: openerp/osv/orm.py [hunk 3]
```diff
@@ -4178,22 +4178,31 @@ def write(self, cr, user, ids, vals, context=None):
...
+                        query_params.append(parent_val)
+                    else:
+                        query_clause += (query_clause and ' OR ' or '') + "%s IS NOT NULL" % (parent_param,)
+            if parents_changed:
```
Comment (rco-odoo) [inline openerp/osv/orm.py]: No, please don't change the type of that variable!
If sensitive fields are in `vals`, then `query_clauses` will be non-empty.

File: openerp/osv/orm.py [hunk 4]
```diff
@@ -4178,22 +4178,31 @@ def write(self, cr, user, ids, vals, context=None):
...
+                query += (query_clause and (' AND (' + query_clause + ')') or '') + " ORDER BY %s" % (self._parent_order,)
```
Comment (rco-odoo) [inline openerp/osv/orm.py]: Please ORDER BY `parent_order`.

**Comment by gurneyalex** [comment]
@rco-odoo done. Thanks for considering this PR.

**Comment by rco-odoo** [comment]
@odony This looks good to me.
Can you have a look at it, too? Thanks!

**Comment by Yenthe666** [comment]
Hello,

Thank you for your contribution but the branch 7.0 is no longer supported. We only support the last 3 stable versions so no longer accepts patches into this branch.
We apology if we could not look at your request in time.
If the contribution still makes sense for the upper version, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

Regards,
Yenthe

**Pull request closed without merging**